### PR TITLE
Add basic tests and make them run with travis

### PR DESCRIPTION
**Pull request by KevinEtienne**
Need to make a decision if we need to drop support for:
 - python2
 - django < 1.8

### CODE REVIEW

**Comment by KevinEtienne** [comment]
Still need a few refactor and see if it's possible to make it compatible easily with django1.9

**Comment by KevinEtienne** [comment]
This PR is slightly bigger than what I originally planned but I'm happy to break it down if there are too much changes. @meshy @Ian-Foote @adam-thomas can you you review please?

File: tests/run.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+
+
+django.setup()
+
+
+# DiscoverRunner requires `django.setup()` to have been called
+from django.test.runner import DiscoverRunner  # noqa
+
+
+class TestRunner(ColourRunnerMixin, DiscoverRunner):
```
Comment (meshy) [inline tests/run.py]: I don't think we need discover runner on django 1.6.

File: rewrite_external_links/middleware.py [hunk 2]
```diff
@@ -21,30 +21,32 @@ class RewriteExternalLinksMiddleware(object):
...
-                a = str('''%(before)s%(root)s?link=%(link)s&next=%(next)s%(after)s''' % {
-                    'root': self.external_link_root,
-                    'next': next,
-                    'before': m.group('before'),
+                return '{before}{root}?link={link}&next={next}{after}'.format(
+                    root=self.external_link_root,
+                    next=next,
+                    before=m.group('before'),
                     # unescape the link before encoding it to ensure entities
                     # such as '&' # don't get double escaped
```
Comment (adam-thomas) [inline rewrite_external_links/middleware.py]: There's a stray `#` in the middle of this line :)

File: tests/test_middlewares.py [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+from rewrite_external_links.middleware import RewriteExternalLinksMiddleware
+
+
+class TestRewriteExternalLinksMiddleware(TestCase):
+    def setUp(self):
+        self.middleware = RewriteExternalLinksMiddleware()
+        self.request = MagicMock()
+        self.content_type = 'text/html'
+        self.link = 'http://example.com'
+        self.content = '<a    href="{}"></a>'.format(self.link).encode()
```
Comment (adam-thomas) [inline tests/test_middlewares.py]: Why the random spaces in `a     href` here? :)

**Comment by adam-thomas** [comment]
LGTM aside from a couple of small nitpicks :)

File: tests/test_middlewares.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+from django.http.response import HttpResponse
```
Comment (Ian-Foote) [inline tests/test_middlewares.py]: I'd call this file `test_middleware.py`.

**Comment by Ian-Foote** [comment]
We should have tests for [IRIs](http://www.w3.org/International/articles/idn-and-iri/).

File: tests/test_middlewares.py (hunk 3 above)
Comment (KevinEtienne) [inline tests/test_middlewares.py]: It's part of what allow the regex. Ideally we should have a couple more tests to show what is possible.

File: tests/run.py (hunk 1 above)
Comment (KevinEtienne) [inline tests/run.py]: This is discover runner from django. Not sure what to replace it with.

Comment (meshy) [inline tests/run.py]: Oops!

**Comment by KevinEtienne** [comment]
@Ian-Foote thanks, nice idea! Added https://github.com/incuna/django-rewrite-external-links/issues/3

File: tests/run.py (hunk 1 above)
Comment (KevinEtienne) [inline tests/run.py]: :)

File: rewrite_external_links/middleware.py [hunk 5]
```diff
@@ -21,30 +21,32 @@ class RewriteExternalLinksMiddleware(object):
...
     """
 
-    #                            <a ... href="                www.xxx.com                                " ... />
-    extlinks = re.compile(r'''(?P<before><a[^>]*href=['"]?)(?P<link>https?://''' + safe_urls + '''[^'">]*)(?P<after>[^>]*)''')
+    # <a ... href="    www.xxx.com                                " ... />
+    extlinks = re.compile(r'''
+        (?P<before><a[^>]*href=['"]?)
+        (?P<link>https?://{}[^'">]*)
+        (?P<after>[^>]*)
+    '''.format(safe_urls), re.VERBOSE)
```
Comment (Ian-Foote) [inline rewrite_external_links/middleware.py]: :+1: for `re.VERBOSE`!

Comment (Ian-Foote) [inline rewrite_external_links/middleware.py]: You can now move the comments inline with the regex:

```python
extlinks = re.compile(r'''
    (?P<before><a[^>]*href=['"]?)  # <a ... href="
    (?P<link>https?://{}[^'">]*)  # www.xxx.com
    (?P<after>[^>]*)  # " ... />
'''.format(safe_urls), re.VERBOSE)
```

I'm not sure the comments are in sync though...

File: rewrite_external_links/middleware.py [hunk 6]
```diff
@@ -16,35 +16,36 @@ class RewriteExternalLinksMiddleware(object):
...
-        <a href="/external_link/?link=http://www.xxx.com&next=...">
+        <a href="/external_link/?link=http://www.example.com&next=...">
     """
 
-    #                            <a ... href="                www.xxx.com                                " ... />
-    extlinks = re.compile(r'''(?P<before><a[^>]*href=['"]?)(?P<link>https?://''' + safe_urls + '''[^'">]*)(?P<after>[^>]*)''')
+    extlinks = re.compile(r'''
+        (?P<before><a[^>]*href=['"]?)  # content from `<a` to `href='`
+        (?P<link>https?://{}[^'">]*)  # href link
+        (?P<after>[^>]*)  # content after the href attribute to the closing bracket `>`
```
Comment (KevinEtienne) [inline rewrite_external_links/middleware.py]: what about this @Ian-Foote

Comment (Ian-Foote) [inline rewrite_external_links/middleware.py]: :+1:

**Pull request merged**